phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexUpgradeToolIT.java

```diff
@@ -0,0 +1,506 @@
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized.Parameters;
+import org.junit.runners.Parameterized;
+
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Properties;
+import java.util.UUID;
+
+@RunWith(Parameterized.class)
+@Category(NeedsOwnMiniClusterTest.class)
+public class IndexUpgradeToolIT extends BaseTest {
+
+    private final boolean mutable;
+    private final boolean upgrade;
+    private StringBuilder optionsBuilder;
+    private String tableDDLOptions;
+
+    private static Map<String, String> serverProps = Maps.newHashMapWithExpectedSize(1), clientProps = Maps.newHashMapWithExpectedSize(1);
+
+    private String [] indexesList = {"TEST.INDEX1", "TEST.INDEX2", "TEST1.INDEX3","TEST1.INDEX2","TEST1.INDEX1","TEST.INDEX3"};
+    private String [] tableList = {"TEST.MOCK1","TEST1.MOCK2","TEST.MOCK3"};
+
+    private String inputList = "TEST.MOCK1:INDEX1,INDEX2;TEST1.MOCK2:INDEX3,INDEX2,INDEX1;TEST.MOCK3:INDEX3";
+
+
+    @Before
+    public void setup () throws Exception {
+        optionsBuilder = new StringBuilder();
+
+        if (upgrade) {
+            if ((Boolean.TRUE.toString()).equals(clientProps.get(QueryServices.INDEX_REGION_OBSERVER_ENABLED_ATTRIB))) {
+                tearDownMiniClusterAsync(1);
+            }
+            clientProps.put(QueryServices.INDEX_REGION_OBSERVER_ENABLED_ATTRIB, Boolean.FALSE.toString());
+        } else {
+            if ((Boolean.FALSE.toString()).equals(clientProps.get(QueryServices.INDEX_REGION_OBSERVER_ENABLED_ATTRIB))) {
+                tearDownMiniClusterAsync(1);
+            }
+            clientProps.put(QueryServices.INDEX_REGION_OBSERVER_ENABLED_ATTRIB, Boolean.TRUE.toString());
+        }
+        if (!mutable) {
+            optionsBuilder.append(" IMMUTABLE_ROWS=true");
+        }
+        setUpTestDriver(new ReadOnlyProps(serverProps.entrySet().iterator()),
+                new ReadOnlyProps(clientProps.entrySet().iterator()));
+        tableDDLOptions = optionsBuilder.toString();
+        try (Connection conn = DriverManager.getConnection(getUrl(), new Properties())) {
+            conn.setAutoCommit(true);
+            prepareSetup(conn);
+        }
+
+    }
+
+
+    private void prepareSetup(Connection conn) throws SQLException {
+        //inputList is "TEST.MOCK1: INDEX1, INDEX2; TEST1.MOCK2: INDEX3, INDEX2, INDEX1; TEST.MOCK3: INDEX3";
+        conn.createStatement().execute("CREATE TABLE TEST.MOCK1 (id bigint NOT NULL PRIMARY KEY, a.name varchar, sal bigint, address varchar)"+tableDDLOptions);
```

Comment:
Yes, I agree. But here, since I needed to test across different schemas, tables, etc., I chose to have a defined list which can be prepopulated to pass in -tb or -f parameter as per the tests.